File: cpp/cpp_100-999/207.cc
// C++
// leetcode 207
// https://leetcode-cn.com/problems/course-schedule/
// 很慢的屑方法

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
    	bool iRelayj[numCourses][numCourses];
    	memset(iRelayj, 0, sizeof(iRelayj));
    	for(auto & p : prerequisites) {
    		iRelayj[p[0]][p[1]] = true;
    	}
    	bool finished[numCourses];
    	memset(finished, 0, sizeof(finished));
    	int done = 0;
    	bool lastTurnLearnedSomething = true;
    	
    	while(done < numCourses && lastTurnLearnedSomething) {
    		lastTurnLearnedSomething = false;
	    	for(int i = 0; i < numCourses; i++) {
	    		// 看一下课i能否学
	    		if(finished[i]) continue;	// 之前学过了
	    		bool canLearn = true;
	    		for(int j = 0; j < numCourses; j++) {
	    			if(iRelayj[i][j] == true) {
	    				canLearn = false;
	    				break;
	    			}
	    		}
	    		if(canLearn) {
	    			// 学课i
	    			lastTurnLearnedSomething = true;
	    			done++;
	    			finished[i] = true;
	    			for(int j = 0; j < numCourses; j++) {
	    				iRelayj[j][i] = false;
	    			}
	    		}
	    	}
	    }
	    return done == numCourses;
    }
};

File: Go/Go_100-999/207.go
// Go
// leetcode 207
// https://leetcode.cn/problems/course-schedule/

func canFinish(numCourses int, prerequisites [][]int) bool {
	after := make([][]int, numCourses)
	preNum := make([]int, numCourses)
	taken := make([]bool, numCourses)
	for _, p := range prerequisites {
		c, before := p[0], p[1]
		preNum[c]++
		after[before] = append(after[before], c)
	}

	cnt := 0
	for cnt < numCourses {
		cur := 0
		for i:=0; i< numCourses; i++ {
			if taken[i] { // 上过
				continue
			}

			if preNum[i] != 0 { // 还要上前置课程
				continue
			}

			// 上i课
			taken[i] = true
			for _, next := range after[i] {
				preNum[next]--
			}
			cur++
		}
		if cur == 0 {
			break
		}
		
		cnt+=cur
	}
	return cnt == numCourses
}